proposals/0303-swiftpm-extensible-build-tools.md

```diff
@@ -286,131 +266,316 @@ protocol TargetBuildContext {
         /// SwiftPM version.
         var moduleName: String { get }
         
-        /// The path of the target source directory.
-        var targetDir: FilePath { get }
+        /// Path of the target source directory.
+        var targetDirectory: Path { get }
+        
+        /// Path of the public headers directory, if any (Clang targets only).
+        var publicHeadersDirectory: Path? { get }
     }
 
-    /// The path of an output directory where the plugin or the build commands
+    /// The path of a writable directory where the plugin or the build commands
     /// it constructs can write anything it wants to. This includes generated
-    /// source files that should be further processed (by specifying them using
-    /// `CommandConstructor.addGeneratedSourceFile()`), and it could include
+    /// source files that should be processed further, and it could include
     /// any caches used by the build tool or by the plugin itself. The plugin
     /// is in complete control of what is written under this directory, and
     /// the contents are preserved between builds.
-    var outputDir: FilePath { get }
+    ///
+    /// A plugin would usually create a separate subdirectory of this directory
+    /// for each command it creates, and the command would be configured to
+    /// write its outputs to that directory. The plugin may also create other
+    /// directories for cache files and other file system content that either
+    /// it or the command will need.
+    var pluginWorkDirectory: Path { get }
+    
+    /// The path at which SwiftPM will write a postbuild result file, if any
+    /// postbuild commands are created. This is a JSON file whose contents are
+    /// defined in SE-0303 and which may be extended by future proposals. The
+    /// file may not be written if no postbuild command ends up being created.
+    /// The plugin may choose to pass this path to a postbuild command on the
+    /// command line or in the environment.
+    var postbuildResultFile: Path { get }
```

Comment:
this is only relevant to post build plugins, right? I wonder if we need a different context object for different type of plugins or find some other way to customize the context API along these lines as we are likely to have more such attributes that are specific to the type of plugin and we would not want to "pollute" the general context with

Reply:
Because we wanted to make the capabilities very generic (combining the prebuild, build, and postbuild) capabilities, there's actually no such thing as a postbuild plugin anymore.  Rather, any plugin with the build capability can produce an arbitrary assortment of prebuild commands, regular build commands, and postbuild commands.

For different capabilities I agree that the input context would be different.  `TargetBuildContext` is very much geared toward build-oriented plugins.

Reply:
Another approach here, which is what I had originally done but then changed it to this, would be to allow the plugin to make up a path and pass it to the CommandConstructor in the creation of a postbuild command.  SwiftPM would then create a result bundle file at that location.  That might also allow us to extend the API in the future to be more fine-grained about what kind of results it wants there, e.g. tests or no tests, etc.

Reply:
That would be an easy change to make, and as I revisit this, I'm inclined to do so.